csrc/parallel_dimension_map.h

```diff
@@ -41,7 +41,19 @@ class ParallelDimensionMap {
     return dim_map_;
   }
 
-  Val* getNumThreadsEachBlock() const;
+  //! Get the number of threads per each CTA used for computation. When there is
+  //! no warp specialization, the result is trivial: it is just the product of
+  //! parallel dimensions of TIDx, TIDy and TIDz. If we do have warp
+  //! specialization, this returns the number of threads used for computing. For
+  //! example, if we have a simple kernel warp specialized on TIDy and all the
+  //! TIDx parallelized IterDomains have extent 32, and all the TIDy
+  //! parallelized IterDomains have extent 16, and there is no TIDz
+  //! parallelization, then we will have:
+  //!   blockDim = (x=32, y=17, z=1)
+  //! And this function will return (32 * 16) because the extra one for TIDy is
```

Comment:
I think that it is less intuitive to the developer that an extra TIDy is added to the CTA than taking one TIDy for circular buffer loads.

e.g., If you wanted to use all 1024 threads, you specify (x=32, y=31, z=1) rather than (x=32, y=32, z=1).

Reply:
If you want to toggle between `warp-specialized`, you'd may have to change the schedule when using all 1024 threads.

Reply:
> I think that it is less intuitive to the developer that an extra TIDy is added to the CTA

This part, I agree. I dislike the fact that the parallel dimension is not attached to an IterDomain but having one mystery +1. I believe this is just a temporary solution, because I can not come up with a better solution that can be implemented in short amount of time. In the future, likely we will be able to slice on parallel types, and select a slice for load and another slice for compute.

> taking one TIDy for circular buffer loads.

How will that work? If you are already using all 1024 threads to do computation, then you can not take any of them out to load.

Reply:
I understand this design choice of adding an extra warp. Since we're cloning loops, we have to schedule the TV from the perspective of the pipelined approach. To get warp specialization, we have to add an extra warp for loads.